test/Where.agda

```diff
@@ -0,0 +1,51 @@
+{-# OPTIONS --no-auto-inline #-}
+module Where where
+
+open import Haskell.Prelude hiding (x; b; n; m)
+
+postulate
+  bool2nat : Bool -> Nat
+{-# COMPILE AGDA2HS bool2nat #-}
+
+-- no outer arguments
+ex1 : Nat
+ex1 = mult num + bool2nat true
+  where
+    num = 42
+
+    mult : Nat -> Nat
+    mult = _* 100
+{-# COMPILE AGDA2HS ex1 #-}
+
+-- nested where
+ex2 : Nat
+ex2 = mult num + bool2nat true
+  where
+    num = 42
+
+    mult : Nat -> Nat
+    mult = _⊗ 100
+      where
+        _⊗_ = _*_
+{-# COMPILE AGDA2HS ex2 #-}
+
+-- with outer arguments
+ex3 : Nat -> Bool -> Nat
+ex3 n b = mult num + bool2nat b
+  where
+    num = 42 + bool2nat b
+
+    mult : Nat -> Nat
+    mult = _* n
+{-# COMPILE AGDA2HS ex3 #-}
+
+-- nested where with outer arguments
+ex4 : Bool -> Nat
+ex4 b = mult 42
+  where
+    mult : Nat -> Nat
+    mult n = bump 5 (bool2nat b)
+      where
+        bump : Nat -> Nat -> Nat
+        bump x y = x * y + n
+{-# COMPILE AGDA2HS ex4 #-}
```

Comment:
Here's a test case:
```agda
-- with pattern matching and multiple clauses
ex5 : List Nat → Nat
ex5 [] = zro
  where
    zro = 0
ex5 (n ∷ ns) = mult num + 1
  where
    num = 42 + ex5 ns

    mult : Nat → Nat
    mult = _* n
{-# COMPILE AGDA2HS ex5 #-}
```

Reply:
Almost there.. I have now come across a rather peculiar behaviour in this compilation:
```agda
ex7 : Nat → Nat
ex7 n₀ = go₁ n₀
  where
    go₁ : Nat → Nat
    go₁ n₁ = go₂ (n₀ ⊗ n₁)
      where
        go₂ : Nat → Nat
        go₂ n₂ = n₀ ⊗ n₁ ⊗ n₂
```
translates to
```hs
ex7 :: Integer -> Integer
ex7 n₀ = go₁ n₀
  where go₁ :: Integer -> Integer
        go₁ n₁ = go₂ (n₀ ⊗ n₁)
          where go₂ :: Integer -> Integer
                go₂ n₂ = n₁ ⊗ n₁ ⊗ n₂
```

After some debugging, the weird DeBruijn index changes when handling clause `go₁ n₁ = ...`, which has to partially apply the inner definition of go₂. While doing that, the indices for the expression `n₀ ⊗ n₁ ⊗ n₂` ~ `@2 @1 @0` change to `@1 @1 @0`. Any guess on what I might be doing wrong here? Here is the relevant part of the log:
```bash
****** Clause ******
(n₁ : Agda.Builtin.Nat.Nat) |-
  n₁@0 = go₂ @0 (Where._⊗_ @1 @0) : Agda.Builtin.Nat.Nat
**********************


 -- Body -- 

Just (go₂ @0 (Where._⊗_ @1 @0))  ~~>  Just (go₂ (Where._⊗_ @1 @0))

 -- Child -- 

Where._._.go₂ :: (n₁ : Agda.Builtin.Nat.Nat) ->
  Agda.Builtin.Nat.Nat -> Agda.Builtin.Nat.Nat
Where._._.go₂ = [in range: /home/omelkonian/git/agda2hs/test/Where.agda:140,9-12 ]
[(n₁ : Agda.Builtin.Nat.Nat) (n₂ : Agda.Builtin.Nat.Nat) |-
   n₁@1 n₂@0 = Where._⊗_ (Where._⊗_ @2 @1) @0 : Agda.Builtin.Nat.Nat]

  ~~>(apply)  

 -- Child -- 

Where._._.go₂ :: Agda.Builtin.Nat.Nat -> Agda.Builtin.Nat.Nat
Where._._.go₂ = [in range: /home/omelkonian/git/agda2hs/test/Where.agda:140,9-12 ]
[(n₂ : Agda.Builtin.Nat.Nat) |-
   n₂@0 = Where._⊗_ (Where._⊗_ @1 @1) @0 : Agda.Builtin.Nat.Nat]

```

Reply:
I think the problem is that `apply`/`applyE` is not actually the right move here. What happens is this:
``(\ db1 db0 -> db2 * db1 * db0) `apply` db1`` --> `db1 * db1 * db0`. Basically, `apply` gets rid of a binder, thus shifting any outside de Bruijn indices down, but what you want in this case is to go inside the binder. So for definitions that would mean dropping the patterns for the outer variables in the clauses, and the corresponding arguments from the type.

Reply:
Finally went with `apply`-ing and reverting all bodies (see `applyNoBodies` helper), kind of hack-ish but works :)